Add FlaxCLIPTextModelWithProjection

pcuenca opened the pull request:
# What does this PR do?

`FlaxCLIPTextModelWithProjection` is necessary to support the Flax port of Stable Diffusion XL: https://huggingface.co/stabilityai/stable-diffusion-xl-refiner-1.0/blob/fb6d705fb518524cabc79c77f13a0e7921bcab3a/text_encoder_2/config.json#L3

I can add some tests, if necessary, after this approach is validated.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

@patrickvonplaten @patil-suraj @sanchit-gandhi @younesbelkada

pcuenca committed:
Add FlaxClipTextModelWithProjection

This is necessary to support the Flax port of Stable Diffusion XL: https://huggingface.co/stabilityai/stable-diffusion-xl-refiner-1.0/blob/fb6d705fb518524cabc79c77f13a0e7921bcab3a/text_encoder_2/config.json#L3

Co-authored-by: Martin Müller <martin.muller.me@gmail.com>
Co-authored-by: Juan Acevedo <juancevedo@gmail.com>

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_25254). All of your documentation changes will be reflected on that endpoint.

patrickvonplaten commented:
Should we maybe for now just add it in a subfolder of sdxl in diffusers here: https://github.com/huggingface/diffusers/tree/main/src/diffusers/pipelines/stable_diffusion_xl  instead of having to rely on `transformers` here? I'm not 100% convinced this model is really needed for core transformers usage.

Would also not force the user to have to install transformers from main :-)

pcuenca commented:
> Should we maybe for now just add it in a subfolder of sdxl in diffusers here: https://github.com/huggingface/diffusers/tree/main/src/diffusers/pipelines/stable_diffusion_xl instead of having to rely on `transformers` here? I'm not 100% convinced this model is really needed for core transformers usage.

The [PyTorch version of the same model was added 9 months ago](https://github.com/huggingface/transformers/blob/bd90cda9a6bb4723515c17df1192e53abc8e36e3/src/transformers/models/clip/modeling_clip.py#L1198), so I assumed it was ok.

But sure, we can do that. In that case, how do we deal with it?
- Change the library to `diffusers` here: https://huggingface.co/stabilityai/stable-diffusion-xl-refiner-1.0/blob/main/model_index.json#L15. Unless I'm mistaken, then we'd need to distribute the flax weights separately, or use a branch.
- Create a hack in diffusers to map the library.

> 
> Would also not force the user to have to install transformers from main :-)

Yes, of course, this was meant as the long-term solution.

patrickvonplaten commented:
Ah yeah good point JAX & PyTorch share the same config - this will become complicated indeed then. Ok let's try to get it merged here. CLIP is important enough to be merged to `transformers` indeed

pcuenca commented on src/transformers/models/clip/modeling_flax_clip.py:
```diff
@@ -1007,6 +1007,79 @@ class FlaxCLIPTextModel(FlaxCLIPTextPreTrainedModel):
+            return (text_embeds, pooled_output) + text_outputs[2:]
+
+        return FlaxBaseModelOutputWithPooling(
+            last_hidden_state=text_embeds,
```
I initially hesitated doing the same, but it makes sense, will fix :)

pcuenca committed:
Use FlaxCLIPTextModelOutput

pcuenca committed:
make fix-copies again

sanchit-gandhi commented on src/transformers/models/clip/modeling_flax_clip.py:
```diff
@@ -155,6 +155,36 @@
+    Base class for text model's outputs that also contains a pooling of the last hidden states.
+
+    Args:
+        text_embeds(`jnp.ndarray` of shape `(batch_size, output_dim`):
```
(nit)
```suggestion
        text_embeds (`jnp.ndarray` of shape `(batch_size, output_dim`):
```

sanchit-gandhi commented on src/transformers/models/clip/modeling_flax_clip.py:
```diff
@@ -1007,6 +1037,79 @@ class FlaxCLIPTextModel(FlaxCLIPTextPreTrainedModel):
+
+    def setup(self):
+        self.text_model = FlaxCLIPTextTransformer(self.config, dtype=self.dtype)
+        self.text_projection = nn.Dense(self.config.projection_dim, use_bias=False)
```
Projection layer should respect the `dtype` of the module
```suggestion
        self.text_projection = nn.Dense(self.config.projection_dim, use_bias=False, dtype=self.dtype)
```

sanchit-gandhi commented on src/transformers/models/clip/modeling_flax_clip.py:
```diff
@@ -1007,6 +1037,79 @@ class FlaxCLIPTextModel(FlaxCLIPTextPreTrainedModel):
+            deterministic=deterministic,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=True,
```
(nit) we usually set `return_dict=return_dict` (as per the arguments to the module) and then handle the outputs by indexing. See the PyTorch equivalent for an example: https://github.com/huggingface/transformers/blob/5ee9693a1c77c617ebc43ef20194b6d3b674318e/src/transformers/models/clip/modeling_clip.py#L1254

sanchit-gandhi commented on src/transformers/models/clip/modeling_flax_clip.py:
````diff
@@ -1007,6 +1037,79 @@ class FlaxCLIPTextModel(FlaxCLIPTextPreTrainedModel):
+    ```python
+    >>> from transformers import AutoTokenizer, FlaxCLIPTextModelWithProjection
+
+    >>> model = FlaxCLIPTextModelWithProjection.from_pretrained("openai/clip-vit-base-patch32")
````
Nice example! Does this model already have Flax weights on the Hub? Or do we need to convert and push them?

pcuenca replied:
It does! :)

sanchit-gandhi reviewed (approved):
Thanks for adding this @pcuenca - looking clean! Good with me to add this Flax model if you require it for SD / SD XL - I'll leave it to you and Patrick to decide on whether it's an appropriate addition, but from a Flax transformers side the design is great. 

Feel free to add the model to the testing file such that it is run by the CI, just need to add the model class here: https://github.com/huggingface/transformers/blob/5ee9693a1c77c617ebc43ef20194b6d3b674318e/tests/models/clip/test_modeling_flax_clip.py#L318

pcuenca committed:
Apply suggestions from code review

Co-authored-by: Sanchit Gandhi <93869735+sanchit-gandhi@users.noreply.github.com>

pcuenca committed:
Use `return_dict` for consistency with other uses.

Co-authored-by: Sanchit Gandhi <93869735+sanchit-gandhi@users.noreply.github.com>

pcuenca committed:
Fix docstring example.

pcuenca committed:
Add new model to FlaxCLIPTextModelTest

pcuenca committed:
Add to IGNORE_NON_AUTO_CONFIGURED list

pcuenca committed:
Merge remote-tracking branch 'origin/main' into flax-clip-text-model-with-projection

pcuenca commented:
Thanks a lot for your in-depth review @sanchit-gandhi! 🙌 I couldn't get back to this PR until today, but I think I addressed all your comments and fixed an error in the example docstring. I was getting some seemingly unrelated CI failures so I just merged the latest `main` to see if they pass.

pcuenca committed:
Fix naming convention.

sanchit-gandhi reviewed (approved):
Looks good - thanks for iterating @pcuenca!

patrickvonplaten reviewed (approved):
Looks good to me!
